Add [Dart] support

dnfield opened the pull request:
Adds Dart support.

Work is primarily based on work done by the Dart SDK team.

Modifications to flat_buffers.dart:
- Add support for Float32 and Int16

Things handled here that don't appear to be handled in SDK version currently:
- Unions
- Non-zero based enums
- Int16 and Float32 values/vectors
- Access to reader functionality for generated types outside of the library
- Namespaces (especially multiple namespaces in a single file)

Things I probably still need but could use some guidance on:
- [x] (Unit? other?) tests
- [x] Documentation (needs update after latest changes)
- [ ] Publish flat_buffers.dart to dartpub

See also #4487

dnfield commented:
Discovered a bug with lists of enums - need to work that out for now

aardappel commented on android/.project:
```diff
@@ -1,20 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Have you tested that this gets correctly re-generated?

dnfield replied:
This wasn't intentional, I had meant to revert this.  Will remove.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
```
Do we have any Dart experts who can help review this file in particular?

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+
+  /// Map containing all strings that have been written so far.  This allows us
+  /// to avoid duplicating strings.
+  Map<String, Offset<String>> _strings = <String, Offset<String>>{};
```
this looks expensive. In C++, string pooling is optional, is that the case here?

dnfield replied:
it is not optional here.  This map is used by `writeString` to achieve string interning in resultant file (and it is only used in write operations obviously).  I'm not sure it's much more expensive than writing duplicates of the string if a string is duplicated. It would not be difficult to allow consumers of `writeString` to opt out of this (or opt in), but I'd think it'd make more sense to opt out.

aardappel replied:
Wether duplicates are likely really depends on the use case. There are plenty of cases where strings don't duplicate at all, and then you end up copying all strings into a map for no reason. FlatBuffers should by default be as low allocation as possible. Since in C++ it is opt-in for that reason, I suggest we do the same in Dart (most other languages have no string pooling feature at this time).

scheglov replied:
I don't have opinion on whether making canonicalization opt-in or opt-out, we generate code that interacts with flat buffers, so we can change it easily.

We store information about Dart packages in flat buffers, and names of classes and methods often duplicates across files in the same package, so we decided to implement this canonicalization.

dnfield replied:
I've made it opt-in to match C++ API

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+/**
+ * List of booleans backed by 8-bit unsigned integers.
+ */
+class _FbBoolList extends Object with ListMixin<bool> implements List<bool> {
```
Why is this kind of special purpose class for bools needed? FlatBuffers vectors (and strings) are typically not compatible with the host language version of these (except in C++ to some extend), so if this creates overhead compared to a more direct way of accessing FlatBuffers data, it may well not be desirable. FlatBuffers is all about avoiding heap allocation, so accessor objects should be as lightweight as possible.

isoos replied:
AFAICT this is actually a dart-y way of creating an efficient FlatBuffer vector.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+ *
+ * The returned unmodifiable lists lazily read values on access.
+ */
+class BoolListReader extends Reader<List<bool>> {
```
why are the ones for bools singled out here?

dnfield replied:
This is a special case where multiple bools are stored as a single uint8.  It's nice for space savings if you have a lot of bools, but it's not interoperable with other langauges and so should probably be removed.

scheglov replied:
Yes, we tried to pack information as much as possible, because we wanted at some point to put it into Dart SDK, but now we don't. Mostly it is for indexes, see https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/summary/idl.dart#L280

It will not be the end of the world (at least not worse than migrating to a slightly different flat buffers implementation) is this goes away.

aardappel replied:
Generating binaries that are 100% compliant with the other implementations is by far the most important concern of all. So yes, please remove this special case. In fact, as part of the test, also reading the binary generated by C++ is a small step in helping to verify that.

dnfield replied:
Fixed this. I like the way they implemented that, unfortunately it's non-standard.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+  /// `0` for `false` and `1` for `true`.
+  void addBool(int field, bool value, [bool def]) {
+    _ensureCurrentVTable();
+    if (value != null && value != def) {
```
bools can be null in Dart??

dnfield replied:
Yes

dnfield replied:
(There is no such thing as a non-nullable type in Dart)

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+/**
+ * The offset from the end of the buffer to a serialized object of the type [T].
+ */
+class Offset<T> {
```
Does Dart guarantee by-value semantics for small wrapper objects like this? In Java for example our offsets are naked ints, to not pay the cost of object allocation everywhere. This sadly of course is less-typesafe.

dnfield replied:
Dart will not treat this as by-val.  Offsets could be made into naked ints, but that seems like a lot of loss in clarity/safety.  Was the Java choice made due to benchmarks demonstrating its perf increases?

aardappel replied:
It was not made based on benchmarks, no. But allocating an object for every offset generated (which there are a lot of) seems excessive.

scheglov replied:
We were told by Dart VM team that allocation is cheap. I can imagine that VM might optimize this to integers. But we have not actually performed benchmarking.

New offset object is created for every object (table), list and string. Is this a lot? Maybe. But probably will not make a lot of performance difference.

aardappel replied:
Allocation is cheap, GC is not. You pay the price elsewhere.

Beyond that, there's the fact that FlatBuffers purpose is to reduce the need for object allocation and copying. So even if Dart's memory management is indeed magically cheaper than other GC languages, it looks weird to have a library like FlatBuffers generate a ton of garbage.

scheglov replied:
I don't have objections against replacing `Offset` with plain `int`.

dnfield replied:
Done.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+  }
+
+  /// Write the given list of 64-bit float [values].
+  Offset writeListFloat64(List<double> values) {
```
Dart appears to have genericity, so why are these functions duplicated per type?

dnfield replied:
This would only make sense as a generic if the methods it called and setup it did could be done generically.  Each function has to do setup that is aware of the size in bytes of the type it's writing, and has to call a concrete non-generic implementation of a writer function.

scheglov replied:
ACK, these functions are similar, but different enough to make it easier and probably faster to have specialized implementations.

aardappel replied:
If its faster then definitely keep them.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+  }
+
+  /// Write the given string [value] and return its [Offset], or `null` if
+  /// the [value] is equal to [def].
```
there's currently no support for string defaults in FlatBuffers, so this can probably be simplified.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+ *
+ * The returned unmodifiable lists lazily read values on access.
+ */
+class Uint32ListReader extends Reader<List<int>> {
```
How and when are these reader classes used? Reading a single field needs to be super fast, so allocating objects is not desirable.

dnfield replied:
There's no allocation involved for these - they're used in ListReaders in the generated code, but only through `const` references.  A `const` class in Dart is a canonicalized/interned instance that requires no allocation beyond the initial static allocation.

aardappel commented on dart/flat_buffers.dart:
```diff
@@ -0,0 +1,1095 @@
+/**
+ * Class that describes the structure of a table.
+ */
+class _VTable {
```
The builder above seems to keep multiple of these around. In most other implementations, there is always at most 1 copy of these tails/offsets around at any one time, and they are reused. To compare vtables, you compare directly against vtables in the buffer, rather than keeping a copy of these lists.

dnfield replied:
I'm pretty sure it's cheaper to allocate new ones than to attempt to reuse and update the object.  Garbage collectors usually do better with shorter lived objects than with having to keep track of/move/reallocate space for long lived objects

aardappel replied:
in this case, you have a list of _VTables that live as long as the builder does. I propose to only let one _VTable object live that long (and inline it into the builder). That should always be an improvement.

dnfield replied:
So looking at the Go implementation, there's this note:

dnfield replied:
Go does something very similar re: storing the previous VTable objects and checking to see if they can be reused

dnfield replied:
C# and Java also both search, but they use the underlying byte data to look for the VTables rather than keeping references.

aardappel commented:
Made some initial comments. And yes, we need a test, docs, and generated code in the repo. Have a look at tests for some other language (e.g. Java) to see what is required.

dnfield commented:
@aardappel  I'll try to work through some of the Dart questions as I have time.  However, much of that code was written by @scheglov (I only added code for reading/writing U/Int16 and Float32s, which was absent).

The bool logic should probably be removed - it's a special case that they appear to use internally but wouldn't be compatible with other platforms.

If @scheglov can comment on any of this it would be helpful, but like I said I'll work through it as I have time as well.

dnfield commented:
I think I'm actually wrong about the way the bools is implemented as I look into it more closely.  It's not packing multiple bools into a single Uint8, it's just converting bools to Uint8 and vice versa. I think this should interop fine with other languages.

aardappel commented:
Yes, would be great to get some feedback from @scheglov

dnfield commented:
I've made the string interning opt-in behavior (defaults to not keeping a map of the strings).  I could work on refactoring the `Offset<T>` class and the `_VTable` usage by the `Builder`, but it might be better to try to come up with some profiling around it first to see.  My suspicion is that the difference will be negligible - I'll see if I can come up with some benchmarks around that.

aardappel commented:
It's hard to benchmark anything related to GC, since if there's no GC happening between the moment you create the garbage and the timing, it will indeed be cheap (most allocations are a pointer increment). It is also hard to take the effect on the cpu cache into account fairly.

On the one hand I am fine with high level languages (compared to C++) having a higher level (and slower API). On the other hand, the reason for FlatBuffers existence (and why people use it) is speed over comfort, so I try to default to that.

dnfield commented:
I was waiting for things to settle on the implementation and code gen.  If that's in good shape now I'll work on those.

dnfield commented:
Nevermind my last comment.  Dart isn't quite as different as I thought.  I'm still working through this, found what I believe would cause a bug in the way vectors of structs are stored and working on that.

dnfield commented:
I didn't have much time for this over the last week - as I'm getting back to it, it seems like the original implementation of this only envisioned supporting tables (not structs).  My effort here so far has been based on that implementation, and all of the logic is wrong when handling a struct (it writes and attempts to read it as a vtable).

So basically, I haven't forgotten about this but it's going to take me a bit more time to rework it to support structs properly.  @aardappel, if you'd like me to close this in the mean time let me know (or just close it yourself I suppose?).

aardappel commented:
@dnfield there is no need to close it, it can stay open as long as you have intentions to finish it.

dnfield commented:
I have the reading part reworked and seems good for a C++ generated flatbuffer.

Still working through the writing portion.

dnfield commented:
At this point I should just be down to finishing up the tests.

I also want to look at restructing the /dart folder a bit to make it match the expected format for Dart pub.

dnfield commented:
I think this should be pretty close.  I've restructed the Dart folder a bit to make it match what dartpub will expect to see.  It can be published using `pub publish` from that folder.

I've added tests in there and the sample in there (with pointers to that folder from the `tests` and `samples` folder shell scripts) - that's a typical setup for pub packages.

I can go ahead and publish it and add you as a publisher as well if you like.

isoos commented:
> I've added tests in there and the sample in there (with pointers to that folder from the tests and samples folder shell scripts) - that's a typical setup for pub packages.

pub.dartlang.org will display the example file if it matches one of the many recognized patterns, e.g. `example/<package_name>.dart`, or `example/example.dart` among others. If you could rename the file it will be picked up automatically.

isoos commented on dart/pubspec.yaml:
```diff
@@ -0,0 +1,18 @@
+documentation: https://google.github.io/flatbuffers/index.html
+dependencies:
+dev_dependencies:
+  test: any
```
It is better to specify version ranges, e.g. here
`test: ^0.12.33` (will be auto-upgraded when 0.12.34 will come out, but won't do it for 0.13.0).

`test_reflective_loader : ^0.1.4`

isoos commented on dart/pubspec.yaml:
```diff
@@ -0,0 +1,18 @@
+- Paul Berry
+homepage: https://github.com/google/flatbuffers
+documentation: https://google.github.io/flatbuffers/index.html
+dependencies:
```
if there are no dependencies, this can be removed

isoos commented on dart/pubspec.lock:
```diff
@@ -0,0 +1,311 @@
+# Generated by pub
```
For published libraries Dart usually recommends to not publish the `pubspec.lock` (and it is usually put in `.gitignore` to keep it out of the repository). IIRC this won't be an issue if you publish it, but you may get a warning about it.

dnfield commented:
Thanks for that @isoos

dnfield commented:
This should be good to go as far as I'm concerned.  @aardappel - would you like me to publish it with you as an author/publisher as well?

aardappel commented on src/idl_gen_dart.cpp:
```diff
@@ -0,0 +1,745 @@
+    object_name = struct_def.name;
+    GenDocComment(struct_def.doc_comment, &code, object_name);
+
+    std::string impl_name = "_" + struct_def.name + "Impl";
```
please use auto where possible.

aardappel commented on dart/example/monster_my_game.sample_generated.dart:
```diff
@@ -0,0 +1,456 @@
+import 'package:flat_buffers/flat_buffers.dart' as fb;
+
+
+class Color {
```
Is this the only way to do enums in Dart? Can't we have integer constants instead somehow (without a class)?

dnfield replied:
Raw enums in dart exists but must be 0 indexed, which is not the case in FlatBuffers.  If we used raw integer constants we'd lose type-safety which doesn't seem like a good tradeoff in this case, but could be possible.  My preference would be to keep the enum-like classes here.

aardappel replied:
It is more important that it matches the semantics of FlatBuffers I feel, which represents them as integers. If enums cause object allocation, that be quite odd. The other languages also prefer speed over convenience whereever applicable.

If you feel strongly about convenience, a seperate "object API" (that sits on top of the base API) can be added.. one exists for C++ and is planned for other languages.

dnfield replied:
I think it'd make sense to refactor this then to use raw ints and have an enum helper class available with some kind of `toInt`/`fromInt` methods/constructors.  I'll work on that.

dnfield replied:
Actually, reexamining the current implementation - it isn't doing all the allocating it seems to at first.  The use of `const` on these means they're all only allocated once: 
```
  static const Color Red = const Color._(0);
  static const Color Green = const Color._(1);
  static const Color Blue = const Color._(2);
```
These only require the initial allocations for each value - so three allocations.  Those three allocations are statically stored in the program and reused - there's no 'new' allocation used anywhere related to these (it's always `const`).  A change to this to use ints would be fairly simple, but it would also be less clear and not really save a whole lot in the end.  Are you still opposed to having it done this way?

aardappel replied:
I guess that is much more acceptable then. If this is idiomatic for doing enums in Dart I guess it is ok.

My thinking is that you should want a representation that matches FlatBuffers semantics (which is cross-language) rather than trying to force the language way of doing things.

aardappel commented on dart/example/monster_my_game.sample_generated.dart:
```diff
@@ -0,0 +1,456 @@
+
+  double _x;
+  @override
+  double get x {
```
Ok, I'm confused about a bunch of things here:

1. Why does there need to be an abstract class and impl, rather than just 1 class?
2. Why `double`? Are these 32bit in Dart?
3. Does the declaration of `double _x` mean an actual variable is stored? If this an accessor, it should really store zero things, as the data is in the buffer.
4. For my education, what is this `??=` doing?

dnfield replied:
1. I did this mainly because it followed the design in the SDK code.  I'm not necessarily committed to this design, but I suppose it could allow for other implementations of the same model class that aren't FlatBuffer driven more easily (e.g. a `MonsterJsonImpl` or some such thing).
2. `double` is the only floating point type available in Dart and is defined by the spec to be 64 bits (int, per the spec, can be as many bits as you have memory).  There is some notion of lists/vectors of Float32 or Float64s, but not individual numbers that I'm aware of.
3. The declaration `double _x;` is equivalent to `double _x = null;`.  Null is a constant in dart, and as far as I understand does not involve any allocation.  However, I'm pretty sure that internally there'd be a pointer generated for the reference to null.  
4. `??=` is an operator that assigns the variable if it is null, or leaves it as is if it is not null.

Basically, for 3 and 4 - it's caching access to the underlying buffer so that subsequent calls to the getter will not have to touch the underlying buffer again.  I don't have any tests around this, but I'd imagine there are cases where it's faster one way or the other.  This is how it was developed in the SDK code and I kept it.

aardappel replied:
1. in what SDK code? I don't think it makes much sense to inherit from automatically generated classes that represent data in a FlatBuffer.
3. With allocated, I meant stored, sorry. So yes, these variables shouldn't exist.
4. So this is attempting to cache all values? I'd think it should just load from the buffer on each occasion.

dnfield replied:
Ok.  I think this should be plenty fast either way.  I'll get rid of the inherited impl classes too

aardappel commented on dart/example/monster_my_game.sample_generated.dart:
```diff
@@ -0,0 +1,456 @@
+  double get z;
+
+  factory Vec3(List<int> bytes) {
+    fb.BufferContext rootRef = new fb.BufferContext.fromBytes(bytes);
```
This appears to be creating a `BufferContext` for each `Vec3`, or? what exactly is this function meant to do? Why are bytes stored in ints?

dnfield replied:
This is if you want to create a `Vec3` object from a list of bytes (the `List<int>` - Dart does not have a `byte` type, int storage is dynamic per the spec, similar to JavaScript).  That factory constructor should not re-allocate the bytes passed in if the list is actually a `Uint8List` from the typed_data library - otherwise it will create a copy of the data into a `Uint8List` for the user.

aardappel replied:
Still not getting it.. where do these bytes come from and what do they represent?

dnfield replied:
If you wanted to create a `Vec3` from a raw array of bytes, this would do it.  But we're not creating `BufferContext`s for each `Vec3` in a monster - we're using the constructor on the `Impl` class, which takes the existing `BufferContext` and an `offset`

dnfield replied:
In other words, internally, the factory constructor for `Vec3` is never used.  It would only be used by an external end user who wanted to have a flatbuffer of just `Vec3`s.

dnfield replied:
I suppose I could drop these for structs, but in this implementation, it's possible to build a "flat buffer" with just Vec3s in it (and no root table).  I believe the C implementation supports that as well.  My preference would be to document this as non-standard behavior rather than remove the ability to do it if that works for you.

aardappel replied:
That creates something that is not a FlatBuffer, and is not something that is supported by most implementations afaik. I'd remove it, as it could confuse people, and there really is no need for it.

dnfield replied:
So should this only appear for the `root_type` or can it appear on any `table` and just not `struct`s?

aardappel replied:
If this is going to be a way to decode whole buffers, then yes, allowing it on all tables is nice to have.

aardappel commented on dart/example/monster_my_game.sample_generated.dart:
```diff
@@ -0,0 +1,456 @@
+}
+
+class Vec3Builder extends fb.BuilderHelper {
+  final double _x;
```
Builder classes in other languages don't actually store anything, they directly write to a buffer. The whole design of FlatBuffer is about avoid object construction / copying, which this ignores.

dnfield replied:
This will store additional references.  See my comments above on `example.dart`

aardappel replied:
Again, as per the above, I don't think any values should be stored. There really is no reason to.

dnfield replied:
I think this makes sense to rework as a simple method then on the `Vec3` class.  This will be a substantial change though.

dnfield replied:
Coding is done, working on upating the docs/tests.

aardappel commented on dart/example/monster_my_game.sample_generated.dart:
```diff
@@ -0,0 +1,456 @@
+  /// Finish building, and store into the [fbBuilder].
+  @override
+  int finish(
+    fb.Builder fbBuilder, { bool internStrings = false }) {
```
The internstrings bool should really be in the builder object that contains the string map?

dnfield replied:
This flag gets passed through to that Builder.  This way, particular tables/objects can choose to intern strings or not.

aardappel replied:
It be fine to determine that globally, I think.

dnfield replied:
I'd like to avoid global variables, and I like the idea of being able to intern strings for some objects but not others...  This allows for that.

dnfield replied:
Got it now, this should be an easy change.

dnfield commented on dart/README.md:
```diff
@@ -0,0 +1,13 @@
+# FlatBuffers for Dart
```
This is the content that would show up on pub.dartlang.org.  I can link it to the specific Dart documentation, but it shouldn't be the exact content of DartUsage.md because that has some doxygen specific stuff that wouldn't render properly on pub.dartlang.org

aardappel replied:
Ok, that's fine, leave it here then.

dnfield commented on dart/example/example.dart:
```diff
@@ -0,0 +1,86 @@
+/*
```
I initially had this in `sample/`, but moved it here.  I'd like it to be included in what's distributed on pub.dartlang.org, and this is a common convention for doing so (and will be more immediately helpful to people who install the package).  Do you want a copy of it also included in `sample/`?

aardappel replied:
I certainly don't want copies. Dart packages can't pull from different directories?

I'd prefer it in `sample/` since there it can share the schema, and it is useful for people to see how FlatBuffers is used in different languages all in one place. Also, generally it is a structure we've used so far, and would be nice to not break for the sake of Dart.

isoos replied:
@aardappel: we are encouraging certain patters, and we have settled on the `example/` directory so far. You could script the publishing, and create a temp copy it from `sample/` just before calling `pub publish`, and remove it afterwards.

aardappel replied:
I'd be fine with a publish script that copies things around, as long as they're in git in one location.

aardappel replied:
My knowledge of Dart is minimal, so I'll leave it to you to judge. That said, we cannot have any duplicated files, and we already have 8 or so languages conforming to a particular pattern where tests and samples are in a particular locations and share particular files/schemas, so it be really good not to have to make an exception for Dart's sake.

dnfield commented on CMakeLists.txt:
```diff
@@ -45,6 +45,7 @@ set(FlatBuffers_Library_SRCS
 set(FlatBuffers_Compiler_SRCS
   ${FlatBuffers_Library_SRCS}
   src/idl_gen_cpp.cpp
+  src/idl_gen_dart.cpp
```
Sorry for being dense but can you point me to the file?  I'm not finding another file where individual .cpp files are referenced besides the `BUILD` file and this one (which I updated both).

aardappel replied:
Ah, never mind, looks like Conan deals with this generically.

dnfield commented on dart/lib/flat_buffers.dart:
```diff
@@ -0,0 +1,1168 @@
+
+  /// Map containing all strings that have been written so far.  This allows us
+  /// to avoid duplicating strings.
+  Map<String, int> _strings = <String, int>{};
```
Making this null would require a check on each call of `writeString` to create it if it's null and `intern` is true.  The trade-off is allocating one of these per instance of the `Builder`, or making that check every time someone wants to intern a string.  Neither one seems especially expensive, but it feels more complicated to me to have to check if a previous call to writeString already newed up the map or not.  If you really like I could move it though.

aardappel replied:
Then why not make the null-check indicate if interning is used, and not use an `intern` var at all?

You'd definitely want just one of these per buffer you're creating.

dnfield replied:
So right now only one gets created per buffer.  I believe I'd be meeting your request here by not initializing this by default and adding something like `_strings ??= <String, int>{};` after checking if `intern` is set to true in `writeString`, which is fine but adds a few more instructions for each call to writeString to save a cheap allocation.

aardappel replied:
I'd get rid of the `intern` variable at all and just check `_strings` to see if the string needs to be pooled.

dnfield replied:
Got it now.  Have the builder constructor take the parameter, and from there determine whether to allocate the map for it or not.  That makes more sense.

dnfield commented on dart/lib/flat_buffers.dart:
```diff
@@ -0,0 +1,1168 @@
+}
+
+/// Class that describes the structure of a table.
+class _VTable {
```
You can't nest classes in dart, and I'd rather avoid adding even more fields/methods to the already large `Builder` class.

dnfield commented on dart/pubspec.yaml:
```diff
@@ -0,0 +1,17 @@
+name: flat_buffers
+version: 1.0.0
```
Do you want the package version for dartpub of this to mirror the flatc version?  It seems to me like they could develop independently and be versioned independently.

aardappel replied:
I have to maintain version numbers for a lot of languages, so it is more meaningful if they are in sync.

dnfield commented on dart/test/flat_buffers_test.dart:
```diff
@@ -0,0 +1,480 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:typed_data';
```
Moving this file makes it more complicated to properly include the package files.  It also means that the test suite would not get distributed through dart pub, which expects these files to be in the test folder here.

Could we add a symlink from `tests/` to this?

aardappel replied:
symlinks are problematic in Windows.

Complicated how? Again I'd expect a packaging tool to be able to pull from different places.

dnfield replied:
I'll create a script to move things around and prep the package.

aardappel replied:
has this happened?

dnfield replied:
For some reason, some views are missing changes to this file, but I did add a "CheckOtherLangaugesData" that compares a C++ generated buffer, and a BuilderTest that is similar to a lot of tests the Go tests run.

dnfield commented on dart/test/flat_buffers_test.dart:
```diff
@@ -0,0 +1,480 @@
+  TestPointImpl createObject(BufferContext object, int offset) {
+    return new TestPointImpl(object, offset);
+  }
+}
```
I'll do some work on that.  These were helpful to validate some changes I made from the SDK code, and are largely based on that work.

dnfield commented on dart/example/example.dart:
```diff
@@ -0,0 +1,86 @@
+void main() {
+  // Create the builder here so we can use it for both weapons and equipped
+  // the actual data will only be written to the buffer once.
+  var axe = new myGame.WeaponBuilder(name: 'Axe', damage: 5);
```
The user could still use the `Builder` class more directly, the way the generated code does.

The "builder helpers" do allocate references to the data that will be held until the helper goes out of scope/is garbage collected.

In a lower level language, this would definitely be a concern and more important to avoid - I'm just not as convinced it makes as much sense to avoid in Dart.  The scale at which this would become a problem in Dart will likely already be a problem for other reasons (in other words, I don't see Dart being the best candidate for processing very large FlatBuffers data, even if it wasn't making any unnecessary allocations).  I also am not convinced that the runtime wouldn't decide to make allocations behind our backs anyway (I've certainly seen that in other garbage collected environments), even though we've made the code more difficult to use by trying to avoid any unnecessary allocations. 

I've basically followed the design that was done in the original SDK work, although in this case I think that it's worth keeping both for ease of use and low likelihood of performance or usability improvements.

aardappel replied:
I agree that it is not likely to be an issue for Dart, but it is more of a style issue. FlatBuffers is about creating and reading buffers in-place, and is about efficiency. Even in Python (probably the slowest language of all) we adhere to this pattern.

dnfield replied:
I'm almost set on this - I've renamed the existing builders to "ObjectBuilder", because I think the syntax/usage will feel more natural and is easier to use for people who don't mind the additional allocations.  But I'm also generating builders closer to what you have in the other languages

dnfield commented on dart/lib/flat_buffers.dart:
```diff
@@ -0,0 +1,1168 @@
+/// Reader of lists of boolean values.
+///
+/// The returned unmodifiable lists lazily read values on access.
+class BoolListReader extends Reader<List<bool>> {
```
Moved these down.

dnfield commented:
I've separated out the previous generated Builders into what I'm now calling "ObjectBuilders" - I believe that's in line with the Object model in other langauges, roughly.  I think these classes will be more accessible for most Dart users, and probably meet many performance/memory requirements.

I've added a new set of Builder classes that are more in line with the existing implementations, as well as some tests for those and a test that ensures we can read a C++ generated file.  This did lead to a couple changes in how enums are handled - I had been assuming enums would not allow gaps, but that's not the case for `bit_flag` enums or for otherwise overridden enums.  

The test cases are still under the `dart/` folder, and the script in `tests/` references that - but I'm not sure there's a nice/clean way for Dart and Pub to get necessary dependencies if we put it all under the `tests/` folder.  I'd welcome input on that...

aardappel commented on dart/.gitignore:
```diff
@@ -0,0 +1,6 @@
+.pub/
```
would prefer this to be part of the root `.gitignore`, that way it is easier to overview what is all being ignored

aardappel commented on dart/LICENSE:
```diff
@@ -0,0 +1,24 @@
+Copyright 2012, the Dart project authors. All rights reserved.
```
Not sure if this file can be here. Google has a requirement for copyright assignment when doing a PR to a Google repo. Since this is from code originally from someone at Google, can we let this fall under the root Apache 2 license? Can you ask @scheglov or whoever you got the base code from?

aardappel replied:
I guess if it comes from Chromium then it will be fine for Google. As far as being included in this folder, it could also be copied from the root by your script, if the same license can be used.

aardappel commented on dart/example/monster.fbs:
```diff
@@ -0,0 +1,33 @@
+// Example IDL file for our monster's schema.
```
for example, this files can't be copied, regardless of what structure you choose for the other files.

aardappel commented:
Other than the license and file locations/copies no major issues at this point. Again, hard to tell since I am no Dart programmer, but we iterate after this PR.

dnfield commented:
I did another pass over the Dart and C++ code, fixed a few missing/minor issues.

I've updated the LICENSE in Dart to try to reflect that the flat_buffers.dart code is based on code from chromium - I'm not married to that formulation but perhaps it works for what you're looking for for this repo.

I added a bash script to help with pub publication - it will copy a couple files, do the publish, and then remove them afterwards to avoid file duplication.  Maybe someone who's more adept with pub and dart packaging than I am can come up with a better way to organize these files at some point, but I haven't been able to figure out a way to keep the files the way you have them and still make Dart find the correct packages to run the tests.

dnfield commented:
One last push here - the sample file schema was actually slightly different (had one extra field), updated the dart example to not use that field anymore.  Also removed what I thought was going to be an optional parameter in the C++ but isn't needed.

dnfield commented:
Should the version be 1.9.0 now?  Or will it be 1.10.0 when Dart is support is released?

aardappel commented:
The version should be `1.9.0`, it will only be bumped when the release actually happens.

aardappel commented on dart/publish.sh:
```diff
@@ -0,0 +1,28 @@
+
+command -v pub >/dev/null 2>&1 || { echo >&2 "Require `pub` but it's not installed.  Aborting."; exit 1; }
+
+cp ../samples/monster.fbs example/
```
I see the test/sample files themselves are still in `dart/` though, are you going to move them as well?

dnfield replied:
They wouldn't be able to execute properly in `samples/` - they wouldn't be able to find the package files (e.g. flat_buffers.dart) correctly without some kind of special logic to rewrite them.

dnfield replied:
(again, at least as far as my knowledge of pub and packaging goes)

aardappel replied:
ah :(

aardappel commented:
Ok, if noone has further comments, I can merge.

dnfield commented:
Thanks for all the time and feedback on this!  Looking forward to being able to use it 😄
